Signal/src/ViewControllers/MediaGalleryViewController.swift

```diff
@@ -0,0 +1,710 @@
+//
+//  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
+//
+
+import Foundation
+
+public enum GalleryDirection {
+    case before, after, around
+}
+
+public struct MediaGalleryItem: Equatable {
+    let logTag = "[MediaGalleryItem]"
+
+    let message: TSMessage
+    let attachmentStream: TSAttachmentStream
+    let galleryDate: GalleryDate
+
+    init(message: TSMessage, attachmentStream: TSAttachmentStream) {
+        self.message = message
+        self.attachmentStream = attachmentStream
+        self.galleryDate = GalleryDate(message: message)
+    }
+
+    var isVideo: Bool {
+        return attachmentStream.isVideo()
+    }
+
+    var thumbnailImage: UIImage {
```

Comment:
essential for memory consumption when we're showing 20-40 media tiles. 

We should also consider using pre-downsampled media for the conversation view as well.

Reply:
Sounds like a solid card.

Reply:
Added!